src/meshTools/AMIInterpolation/AMIInterpolation/AMIMethod/directAMI/directAMI.C

```diff
@@ -232,96 +403,107 @@ void Foam::directAMI<SourcePatch, TargetPatch>::calculate
     label tgtFaceI
 )
 {
-    bool ok =
-        this->initialise
-        (
-            srcAddress,
-            srcWeights,
-            tgtAddress,
-            tgtWeights,
-            srcFaceI,
-            tgtFaceI
-        );
+    this->checkPatches();
+
+    // set initial sizes for weights and addressing
+    srcAddress.setSize(this->srcPatch_.size());
+    srcWeights.setSize(this->srcPatch_.size());
+    tgtAddress.setSize(this->tgtPatch_.size());
+    tgtWeights.setSize(this->tgtPatch_.size());
 
-    if (!ok)
+    // check that patch sizes are valid
+    if (!this->srcPatch_.size())
     {
         return;
     }
+    else if (!this->tgtPatch_.size())
+    {
+        WarningIn
+            (
+                "void Foam::AMIMethod<SourcePatch, TargetPatch>::initialise"
+                    "("
+                    "labelListList&, "
+                    "scalarListList&, "
+                    "labelListList&, "
+                    "scalarListList&, "
+                    "label&, "
+                    "label&"
+                    ")"
+            )
+        << this->srcPatch_.size() << " source faces but no target faces" << endl;
 
+        return;
+    }
 
-    // temporary storage for addressing and weights
-    List<DynamicList<label> > srcAddr(this->srcPatch_.size());
-    List<DynamicList<label> > tgtAddr(this->tgtPatch_.size());
+    const pointgpuField& tgtPoints = this->tgtPatch_.getPoints();
+    const pointgpuField& srcPoints = this->srcPatch_.getPoints();
 
+    const faceDatagpuList& tgtFaces = this->tgtPatch_.getFaces();
+    const faceDatagpuList& srcFaces = this->srcPatch_.getFaces();
 
-    // construct weights and addressing
-    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    vectorgpuField srcCf(srcFaces.size());
 
-    // list of faces currently visited for srcFaceI to avoid multiple hits
-    DynamicList<label> srcSeeds(10);
+    thrust::transform
+    (
+        srcFaces.begin(),
+        srcFaces.end(),
+        srcCf.begin(),
+        faceCentreFunctor
+        (
+            this->srcPatch_.getFaceNodes().data(),
+            srcPoints.data()
+        )
+    );
 
-    // list to keep track of tgt faces used to seed src faces
-    labelList srcTgtSeed(srcAddr.size(), -1);
-    srcTgtSeed[srcFaceI] = tgtFaceI;
+    labelgpuList srcCount(this->srcPatch_.size(), 0);
+    labelgpuList tgtCount(this->tgtPatch_.size(), 0);
 
-    // list to keep track of whether src face can be mapped
-    // 1 = mapped, 0 = untested, -1 = cannot map
-    labelList mapFlag(srcAddr.size(), 0);
+    gpuList<labelList> srcAddr(this->srcPatch_.size(), labelList(15, -1));
+    gpuList<labelList> tgtAddr(this->tgtPatch_.size(), labelList(15, -1));
```

Comment:
Yes, I had to learn it after hours of debugging. I flattened the array as a work around. I'll update the PR.